hpy/universal/src/hpymodule.c

```diff
@@ -115,49 +115,174 @@ static PyObject *load_from_spec(PyObject *self, PyObject *spec)
         goto error;
     }
 
-    HPy mod = ((InitFuncPtr)initfn)(&global_ctx);
-    if (HPy_IsNull(mod))
+    HPyContext ctx = get_context(debug);
+    if (ctx == NULL)
+        goto error;
+    HPy h_mod = ((InitFuncPtr)initfn)(ctx);
+    if (HPy_IsNull(h_mod))
         goto error;
-    PyObject *py_mod = _h2py(mod);
-    // XXX close the handle
+    PyObject *py_mod = HPy_AsPyObject(ctx, h_mod);
+    HPy_Close(ctx, h_mod);
 
-    Py_DECREF(name_unicode);
     Py_XDECREF(name);
-    Py_XDECREF(path);
     Py_XDECREF(pathbytes);
     return py_mod;
 error:
-    Py_DECREF(name_unicode);
     Py_XDECREF(name);
-    Py_XDECREF(path);
     Py_XDECREF(pathbytes);
     return NULL;
 }
 
+static PyObject *load(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    static char *kwlist[] = {"name", "path", "debug", NULL};
+    PyObject *name_unicode;
+    PyObject *path;
+    int debug = 0;
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO|p", kwlist,
+                                     &name_unicode, &path, &debug)) {
+        return NULL;
+    }
+    return do_load(name_unicode, path, debug);
+}
+
 static PyObject *get_version(PyObject *self, PyObject *ignored)
 {
     return Py_BuildValue("ss", HPY_VERSION, HPY_GIT_REVISION);
 }
 
 static PyMethodDef HPyMethods[] = {
     {"set_debug", (PyCFunction)set_debug, METH_O, "TODO"},
-    {"load_from_spec", (PyCFunction)load_from_spec, METH_O, "Load a .hpy.so"},
+    {"load", (PyCFunction)load, METH_VARARGS | METH_KEYWORDS, "Load a .hpy.so"},
     {"get_version", (PyCFunction)get_version, METH_NOARGS, "Return a tuple ('version', 'git revision')"},
     {NULL, NULL, 0, NULL}
 };
 
+static int exec_module(PyObject *mod);
+static PyModuleDef_Slot hpymodule_slots[] = {
+    {Py_mod_exec, exec_module},
+    {0, NULL},
+};
 
 static struct PyModuleDef hpydef = {
     PyModuleDef_HEAD_INIT,
     .m_name = "hpy.universal",
     .m_doc = "HPy universal runtime for CPython",
     .m_size = 0,
     .m_methods = HPyMethods,
+    .m_slots = hpymodule_slots,
 };
 
 
+// module initialization function
+int exec_module(PyObject* mod) {
+    HPyContext ctx  = &g_universal_ctx;
+    HPy h_debug_mod = HPyInit__debug(ctx);
+    if (HPy_IsNull(h_debug_mod))
+        return -1;
+    PyObject *_debug_mod = HPy_AsPyObject(ctx, h_debug_mod);
+    HPy_Close(ctx, h_debug_mod);
+
+    if (PyModule_AddObject(mod, "_debug", _debug_mod) < 0)
+        return -1;
+
+    return 0;
+}
+
+static void init_universal_ctx(HPyContext ctx)
+{
+    if (!HPy_IsNull(ctx->h_None))
+        // already initialized
+        return;
+
+    // XXX this code is basically the same as found in cpython/hpy.h. We
+    // should probably share and/or autogenerate both versions
+    /* Constants */
+    ctx->h_None = _py2h(Py_None);
```

Comment:
I'm not sure I understand why the ctx initialization is here inside hpymodule. Could you explain a bit?

Reply:
before this PR the fields were statically initialized to things like `CONSTANT_H_NONE`, `CONSTANT_H_TRUE` etc, which were small integers defined in an enum. And the old `handles.c` made sure to allocate the corresponding handle at the "right place". So this code is the equivalent of what was inside the old `allocate_more_handles`.

It would be nice to statically initialize things like this:
```c
struct _HPyContext_s g_universal_ctx = {
    ...,
    h_None = _py2h(Py_None),
```

but I don't think it is legal to call `_py2h` in a static initializer